qiskit_experiments/data_processing/nodes.py

```diff
@@ -0,0 +1,204 @@
+# that they have been altered from the originals.
+
+"""Different data analysis steps."""
+
+from abc import abstractmethod
+from typing import Any, Dict, Optional
+import numpy as np
+
+from qiskit_experiments.data_processing.base import DataAction
+
+
+class Kernel(DataAction):
+    """User provided kernel."""
+
+    def __init__(self, kernel_, name: Optional[str] = None):
+        """
+        Args:
+            kernel_: Kernel to kernel the data.
+            name: Optional name for the node.
+        """
+        self.kernel = kernel_
+        self.name = name
+        super().__init__()
+        self._accepted_inputs = ["memory"]
+
+    @property
+    def node_output(self) -> str:
+        """Key under which Kernel stores the data."""
+        return "memory"
+
+    def process(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Args:
+            data: The data dictionary to process.
+
+        Returns:
+            processed data: A dict with the data stored under "memory".
+
+        Raises:
+            DataProcessorError: if the data has no memory.
+        """
+        return {self.node_output: self.kernel.kernel(np.array(data["memory"]))}
+
+
+class Discriminator(DataAction):
+    """Backend system discriminator."""
+
+    def __init__(self, discriminator_, name: Optional[str] = None):
+        """
+        Args:
+            discriminator_: The discriminator used to transform the data to counts.
+                For example, transform IQ data to counts.
+            name: Optional name for the node.
+        """
+        self.discriminator = discriminator_
+        self.name = name
+        super().__init__()
+        self._accepted_inputs = ["memory", "memory_real", "memory_imag"]
+
+    @property
+    def node_output(self) -> str:
+        """Key under which Discriminator stores the data."""
+        return "counts"
+
+    def process(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Discriminate the data to transform it into counts.
+
+        Args:
+            data: The data in a format that can be understood by the discriminator.
+
+        Returns:
+            processed data: A dict with the data stored under "counts".
+
+        Raises:
+            DataProcessorError: if the data does not contain memory.
+        """
+        return {self.node_output: self.discriminator.discriminate(np.array(data["memory"]))}
+
+
+class IQPart(DataAction):
+    """Abstract class for IQ data post-processing."""
+
+    def __init__(self, scale: Optional[float] = 1.0, average: bool = False):
+        """
+        Args:
+            scale: scale by which to multiply the real part of the data.
+            average: if True the single-shots are averaged.
+        """
+        self.scale = scale
+        self.average = average
+        super().__init__()
+        self._accepted_inputs = ["memory"]
+
+    @abstractmethod
+    def _index(self) -> int:
```

Comment:
Perhaps this is abstract method that describes formatting for single complex data point, which is called by the `.process`? i.e.
```python
   @abstractmethod
   def _process(self, point):
      return point.real
```
for IQReal. We may want to rotate the IQ plane using calibrated `self.angle` with this helper method. I think in the result a complex value is no longer a list, i.e. it is converted into complex value. We need to check.

Reply:
Good idea. Introducing `_process` will also make it easy to implement arbitrary rotations. In a follow-up PR we can introduce an IQ processing node that will project the IQ point onto an axis specified by an angle. See e32341e where `_index` was removed and replaced by `_process`.